### PR TITLE
31 audio component

### LINKED ISSUE #31: Build Audio Component
## Acceptance Criteria

- [ ] It should play
- [ ] It should pause
- [ ] It should stop
- [ ] It should handle next and previous tracks
- [ ] It should handle seeking

## Context

Given I want to listen to music when I use the Audio component then common audio functions are available for the user to interact with the audio and the component is cross-platform

### PR DESCRIPTION
**Pull request by Lepozepo**
Fixes #31 

## Creator

- [x] Build relevant tests if any apply
- [x] Ensure there are no new warnings
- [x] Upload an animated GIF showcasing the solution (if it applies)
- [x] Set a relevant reviewer

## Reviewer

- [ ] Run `test` and make sure all tests pass
- [ ] Review function and form on iOS
- [ ] Review function and form on Android
- [ ] Review function and form on Web
- [ ] Review new test logic

### CODE REVIEW

**Comment by Lepozepo** [comment]
screen shot 2017-10-23 at 10 50 29 am

**Comment by conrad-vanl** [comment]
Exciting! I'm not sure how to review this yet. The base acceptance criteria in #31 looks to be fulfilled, but I could see this player needing to work a lot differently in its final state. A few thoughts:

- Audio probably needs to be a global state. As in, there's only _one_ audio source playing ever
- But, there might be multiple instances of player controls: example in their app there's a minimized player view and a full-screen player view
- Global handling of player queues (ie...tapping on a song somewhere in the app tells the audio player to play that song)

For these reasons, I could see wanting to store audio player state in redux: What audio file should be playing, whether the player state is paused or playing, etc.

I had a similar scenario in another project, and the way we solved it (not saying this is correct here, just throwing out ideas):

We had a global `AudioRuntime` component that was rendered out at the root of the app. It didn't render any UI, but it just read from redux state and played sounds based on redux state.

We had different types of player ui components that would interact with the redux state, that essentially controlled the audio playing in the root-level `AudioRuntime` component. Such as a `MiniPlayer` that was fixed to the bottom of the screen, as well as full-screen audio-player controls.

Tapping on a song somewhere in the app fired a redux action that just added that song to the redux state.

**Comment by Lepozepo** [comment]
@conrad-vanl you can review in storybook ^_^. Also I agree we should have some form of control so that songs can't play on top of each other but I'm not sure if that is part of the scope of the component. Once we start building the actual player we can build the "only play one sound at a time" logic to redux. Keeping this component away from business logic could be useful down the road if for any reason we want to be able to play sounds on top of each other.

**Comment by Lepozepo** [comment]
I'm going to refine it a bit more based on the conversation we had :D

**Comment by mary-pr-poppins** [comment]
:shipit: This branch has been deployed to:
  exp://exp.host/@conrad-vanl/apollos-31-audio-component

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

**Comment by mary-pr-poppins** [comment]
:shipit: This branch has been deployed to:
  exp://exp.host/@conrad-vanl/apollos-31-audio-component

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

File: src/@modules/Audio/AudioPause.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import React, { PureComponent } from 'react';
+import getContext from 'recompose/getContext';
+import compose from 'recompose/compose';
```
Comment (conrad-vanl) [inline src/@modules/Audio/AudioPause.js]: I think these can be changed to 

```
import { getContext, compose } from 'recompose';
```

Not a big deal, just syntactic sugar

File: src/@modules/Audio/index.js [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    onPlaybackReachedEnd() {},
+    onPlay() {},
+    onPause() {},
+    onStop() {},
+    onSeek() {},
+    onSeeking() {},
+    children: null,
+  };
+
+  static childContextTypes = {
```
[the remark this answers is not in the document]
Comment (Lepozepo) [reply src/@modules/Audio/index.js]: We use context to couple the components together. Audio components are not meant to be used outside of the scope of Audio.

File: src/@utils/onlyPropTypes.js [hunk 3]
```diff
@@ -0,0 +1,10 @@
+/* eslint-disable react/prop-types */
+import React from 'react';
+import pick from 'lodash/pick';
+
+export default Component => props => (
```
[the remark this answers is not in the document]
Comment (Lepozepo) [reply src/@utils/onlyPropTypes.js]: It doesn't do the same thing and it doesn't quite work the same way. `onlyUpdateForPropTypes` has to be coupled with `setPropTypes`, the extra props will still flow down to the component though, they just won't update. The problem with this is specifically with the way react dom works now, where all extra props are passed down to the DOM element as attributes (which ends up producing a bunch of warnings when we pass camel-case props). This one more easily taps into the way we're dealing with props and removes them.

File: src/@modules/Audio/AudioPause.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (Lepozepo) [reply src/@modules/Audio/AudioPause.js]: Yeah, I've actually wondered whether importing and deconstructing is better than importing the file directly. Does `require` + `babel` still only import that specific bit of code? Or does it bring everything in and deconstruct? I'm all for the short-hand for sure though, just thought you might have something against it >_<

**Pull request merged**